### PR TITLE
Preview size (changes to updatePanel function)

### LINKED ISSUE #2: viewBox top/left are ignored (assumed to be 0)
Open "live_development_sprites.svg" in the Brackets source.
Result: the lightning bolts are all cut off on the left.  Compare to what it looks like if viewer in a browser tab.

Edit the first two numbers in the viewBox attribute.
Result: nothing change in the preview.

This is because updatePanel() only looks at the width/height values in viewBox, not the offset x/y values.

### LINKED ISSUE #8: Viewer is too small when svg height and width are defined as cm units
**OS:** Mac OSX Mavericks

**Brackets Version:** sprint 38 experimental build 0.38.0-12606 (release 4df8afdad)

**Extension Version:** 1.1.2

**Repro  Steps:**
  1. Open a SVG file with the following content (a W3C spec example): https://gist.github.com/lkcampbell/f12f06746602a83cc8af

**Expected Results:**
Should be able to see the entire SVG file.

**Observed Results:**
The height and width of the SVG viewer are too small to see the full image.

**Other Notes:**
Changing the height and width to bigger numbers, with no units or pixel units, makes the SVG viewer larger.

### LINKED ISSUE #12: No preview if SVG doesn't have height/width or viewBox attributes
There should be a default image size (e.g., the HTML default of 300x150) or a way to directly set the preview image size when the SVG does not have any intrinsic dimensions (height and width or viewBox).

### PR DESCRIPTION
**Pull request by AmeliaBR**
This will fix #2, fix #8 and fix #12

The preview size will now:
- Prioritize width/height over viewBox (as it would when displayed).
- Support most units for height/width (it assumes 16px font size and reasonable ex/ch values), although not percentage/viewport units (I couldn't think of reasonable assumptions for conversion to px).
- Scale according to aspect ratio if you have a viewBox plus one of height or width.
- Avoid changing the viewBox attribute if it exists (so left and top offsets are preserved).

(P.S.  You only want the edits to main.js.  I changed the Package info just so I could keep track which version I had installed!)

### CODE REVIEW

**Comment by AmeliaBR** [comment]
Pinging @peterflynn.  Any development plans that would prevent you from adding this fix for three open issues?

File: main.js [hunk 1]
```diff
@@ -79,26 +94,64 @@ define(function (require, exports, module) {
...
+        svgHeight = svgHeight || 150;
+        
         
         // Specify actual width consistent with zoom factor
         var viewWidth = svgWidth * currentState.zoomFactor;
         var viewHeight = svgHeight * currentState.zoomFactor;
         
         // jQ auto lowercases the attr name, making it ignored (http://bugs.jquery.com/ticket/11166 resolved wontfix: "we don't support SVG")
-        $svgRoot[0].setAttribute("viewBox", "0 0 " + svgWidth + " " + svgHeight);
+        if (!viewBoxAttr) {
```
Comment (peterflynn) [inline main.js]: We need to set a viewBox all the time in order to control zooming

Comment (peterflynn) [inline main.js]: Sorry, never mind -- this is correct as-is.

I was tripped by the case where viewBox _and_ width/height are already specified in the original SVG, applying an initial scale/zoom factor out of the box.  It's a bit ambiguous what the zoom feature here should call "100%": the SVG size given by the width/height, or the "natural" size if only the viewBox attribute were present without any scaling due to the additional width/height set.  Currently, it calls the specified width/height (with scaling included) 100%, which seems fair to me.

File: main.js [hunk 2]
```diff
@@ -79,26 +94,64 @@ define(function (require, exports, module) {
...
+                            : bounds[2];
+            }
+            if (!svgHeight) {
+                svgHeight = svgWidth ? svgWidth * bounds[3] / bounds[2]
+                            : bounds[3];
+            }
+        } 
+        // If either has not been defined, or is zero, set defaults:
+        svgWidth  = svgWidth  || 300;   
+        svgHeight = svgHeight || 150;
```
Comment (peterflynn) [inline main.js]: Chrome & Firefox both seem to use 200px in both directions, so it's probably better to just match that.

File: main.js [hunk 3]
```diff
@@ -79,26 +94,64 @@ define(function (require, exports, module) {
...
+        if (viewBoxAttr) {
+            var bounds = viewBoxAttr.split(/[ ,]+/).map(parseFloat);
+            if (!svgWidth) {
+                svgWidth = svgHeight ? svgHeight * bounds[2] / bounds[3]
+                            : bounds[2];
+            }
+            if (!svgHeight) {
+                svgHeight = svgWidth ? svgWidth * bounds[3] / bounds[2]
+                            : bounds[3];
+            }
```
Comment (peterflynn) [inline main.js]: Hmm, Chrome & Firefox differ on what they do when one of width/height is specified along with a viewbox.  Chrome seems to sets the missing one to 100% and scales the _viewBox_ to preserve aspect ratio (potentially revealing parts of the image that were meant to be clipped).  Firefox seems to do what this code does, which seems more sane to me...  So I guess that is indeed the better precedent to follow.

**Comment by peterflynn** [comment]
@AmeliaBR This is really great, thanks a ton!  I merged this into master via 3f95a24a, with some tweaks added in 1ee97a03.  (I had to `git cherry-pick` this since the PR includes other changes specific to your fork that I didn't want to merge upstream -- but the first commit should still show as authored by you).

**Comment by AmeliaBR** [comment]
Thanks @peterflynn.  Glad to see this.

**Comment by peterflynn** [comment]
Thanks again for contributing!  Sorry it took so long before merging it :-/

**Pull request closed without merging**